crates/accelerate/src/sparse_pauli_op.rs

```diff
@@ -334,163 +334,465 @@ pub fn decompose_dense(
+        return;
+    }
+
+    // This set of `extend` calls is effectively an 8-fold unrolling of the "natural" loop through
+    // each bit, where the initial `if` statements are handling the remainder (the up-to 7
+    // least-significant bits).  In practice, it's probably unlikely that people are decomposing
+    // 16q+ operators, since that's a pretty huge matrix already.
+    //
+    // The 8-fold loop unrolling is because going bit-by-bit all the way would be dominated by loop
+    // and bitwise-operation overhead.
+
+    if out.num_qubits & 1 == 1 {
+        out.x.push(x & 1 == 1);
+        out.z.push(z & 1 == 1);
+        x >>= 1;
+        z >>= 1;
+    }
+    if out.num_qubits & 2 == 2 {
+        out.x.extend(&PAULI_LOOKUP_2[x & 0b11]);
+        out.z.extend(&PAULI_LOOKUP_2[z & 0b11]);
+        x >>= 2;
+        z >>= 2;
+    }
+    if out.num_qubits & 4 == 4 {
+        out.x.extend(&PAULI_LOOKUP_4[x & 0b1111]);
+        out.z.extend(&PAULI_LOOKUP_4[z & 0b1111]);
+        x >>= 4;
+        z >>= 4;
+    }
+    for _ in 0..(out.num_qubits / 8) {
+        out.x.extend(&PAULI_LOOKUP_8[x & 0b1111_1111]);
+        out.z.extend(&PAULI_LOOKUP_8[z & 0b1111_1111]);
+        x >>= 8;
+        z >>= 8;
+    }
+
+    let phase = phase % 4;
+    let value = match phase {
+        0 => Complex64::new(out.scale, 0.0) * value,
+        1 => Complex64::new(0.0, out.scale) * value,
+        2 => Complex64::new(-out.scale, 0.0) * value,
+        3 => Complex64::new(0.0, -out.scale) * value,
+        _ => unreachable!("'x % 4' has only four possible values"),
     };
-    let mid = 1usize << (num_qubits - 1);
-    recurse_if_nonzero(
-        Pauli::I,
-        0.5 * factor,
-        &block.slice(s![..mid, ..mid]) + &block.slice(s![mid.., mid..]),
-    );
-    recurse_if_nonzero(
-        Pauli::X,
-        0.5 * factor,
-        &block.slice(s![..mid, mid..]) + &block.slice(s![mid.., ..mid]),
-    );
-    recurse_if_nonzero(
-        Pauli::Y,
-        0.5 * Complex64::i() * factor,
-        &block.slice(s![..mid, mid..]) - &block.slice(s![mid.., ..mid]),
-    );
-    recurse_if_nonzero(
-        Pauli::Z,
-        0.5 * factor,
-        &block.slice(s![..mid, ..mid]) - &block.slice(s![mid.., mid..]),
-    );
+    out.phases.push(phase);
+    out.coeffs.push(value);
+}
+
+/// Pack the information about which row, column and qubit we're considering into a single `usize`.
+/// Complex64 data is 16 bytes long and the operators are square and must be addressable in memory,
+/// so the row and column are hardware limited to be of width `usize::BITS / 2 - 2` each.  However,
+/// we don't need to store at a granularity of 1, because the last 2x2 block we handle manually, so
+/// we can remove an extra least significant bit from the row and column.  Regardless of the width
+/// of `usize`, we can therefore track the state for up to 64 qubits losslessly, which is greater
+/// than the maximum addressable memory on a 64-bit system.
```

Comment:
I'm fairly sure I meant we can track the state for up to 30 qubits losslessly, but just typed "64" twice.  Might come back and think again about it tomorrow.

```suggestion
/// of `usize`, we can therefore track the state for up to 30 qubits losslessly, which is greater
/// than the maximum addressable memory on a 64-bit system.
```